moderation/topwarnings.go

```diff
@@ -0,0 +1,70 @@
+package moderation
+
+import (
+	"database/sql"
+
+	"github.com/jonas747/discordgo"
+	"github.com/jonas747/dstate"
+	"github.com/jonas747/yagpdb/bot"
+	"github.com/jonas747/yagpdb/common"
+)
+
+type WarnRankEntry struct {
+	Rank      int    `json:"rank"`
+	UserID    int64  `json:"user_id"`
+	Username  string `json:"username"`
+	WarnCount int64  `json:"warn_count"`
+}
+
+func TopWarns(guildID int64, offset, limit int) ([]*WarnRankEntry, error) {
+	const query = `SELECT rank, warn_count, user_id FROM
+	(
+		SELECT RANK() OVER (ORDER BY count(message) DESC) AS rank, count(message) as warn_count, user_id
+		FROM moderation_warnings WHERE guild_id = $1 group by user_id
+	) AS warns
+	ORDER BY warn_count desc
+	LIMIT $2 OFFSET $3`
+
+	rows, err := common.PQ.Query(query, guildID, limit, offset)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return []*WarnRankEntry{}, nil
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	result := make([]*WarnRankEntry, 0, limit)
+	for rows.Next() {
+		var member []*discordgo.Member
+		var rank int
+		var tmp []*dstate.MemberState
+		var userID int64
+		var warncount int64
+		var err = rows.Scan(&rank, &warncount, &userID)
+		if err != nil {
+			return nil, err
+		}
+
+		tmp, err = bot.GetMembers(guildID, userID)

```

Comment:
Maybe use GetUsers instead since that works even if they leave

Reply:
maybe it's more reasonable to show ranking only for active members in guild

Reply:
updated for further ideas